migration_design.md

```diff
@@ -0,0 +1,175 @@
+
+Description                             | Shorthand
+--------------------------------------  | -------------------------
+Class deleted                           | A-
+Class added                             | A+
+Class renamed                           | A -> B
+Multiple classes with diff. types       | A{int,str} -> B{int,str} and C{bool} -> D{bool}
+Mult. classes with same type            | A{int} -> B{int} and C{int} -> D{int}
+
+
+**Field changes**
+
+Description                             | Shorthand
+--------------------------------------  | -------------------------
+Rename                                  | {x} -> {y}
+Delete                                  | {x} -> {}
+Type switch (many variants)             | {x:int} -> {x:str}
+Merge (many variants and operators)     | {x:int, y:str} -> {z:x+y}
+Split (many variants and operators)     | {x:int} -> {y:int, z:int} (x -> y = x/2, z = x/3, x-)
+Rename mult., diff. types               | {x:int, y:str} -> {z:int,w:str}
+Rename mult., same. types               | {x:int, y:int} -> {z:int,w:int}
+Move to other class                     | A{x,y}, B{z} -> A{y}, B{x,z}
+Delete + rename to same name            | {x,y} -> {x} (x-, y->x)
+
+
+**Mixing them**
+
+BOOM! Combinatorial explosion!
+
+
+## Types of migrations
+
+### Object migrations
+
+- **Create**: Create a new object
+- **Delete**: Delete existing object
+- **Rename**: Rename an object
+
+### Field migrations
+
+- **Add**: Add a new field
+- **Delete**: Delete a existing field
+- **Rename**: Rename a existing field
+- **Change type**: Change the type of an existing field, ie. from int to string
+- **Merge**: Merge two fields by some rules
+- **Move**: Move to another class (eg. when we introduce subclasses)
+
+### Other features
+
+- **Enumerate existing objects together with new Schema**: Very powerful feature allowing almost any kind of data manipulation
+- **Able to create new objects during migration**
+
+
+## Challenges
+
+- Wording: Should we use Spec or Schema, eg. Spec is closer to class defintions and eg. JavaWriter uses that, Cocoa uses Schema that hints at SQL databases, do we want that?
+- Rules for a magic "magic mode" migration needs to be simple and easy to understand. To much magic makes it hard to
+  reason about behavior
+- How to expose both old and new object model, since the old one no longer exists except as core tables?
+- How much should Magic mode handle?
+- Should "magic mode" and custom migration be mixed, gut feeling says not. It will be too hard for user to reason about
+  their code.
+- Most migration logic should be moved to C++ in a ObjectStoreHelper.cpp or similar to avoid code duplication with Cocoa.
+  Ideally this abstraction layer should have its own independent release cycle from core, ie. Bindings depend on that
+  instead of directly on core.
+
+## Magic mode migration
+
+Magic mode automatically handles the following cases
+
+*Easy to detect*
+
+- New Object
+- Delete object
+- New field
```

Comment:
I think that this is only easy in the trivial case. If a field is renamed, it could look like "delete field" and "new field".